src/crimson/osd/watch.cc

```diff
@@ -123,13 +123,16 @@ seastar::future<> Watch::notify_ack(
   const ceph::bufferlist& reply_bl)
 {
   logger().info("{}", __func__);
-  return seastar::do_for_each(in_progress_notifies,
-    [this_shared=shared_from_this(), reply_bl] (auto notify) {
-      return notify->complete_watcher(this_shared, reply_bl);
-    }
-  ).then([this] {
-    in_progress_notifies.clear();
-    return seastar::now();
+  return seastar::do_with(shared_from_this(),
+    [this, reply_bl] (auto& this_shared) {
+    return seastar::do_for_each(in_progress_notifies,
+      [&this_shared, reply_bl] (auto notify) {
+        return notify->complete_watcher(this_shared, reply_bl);
+      }
+    ).then([this] {
```

Comment:
Do we need `do_with()` and the extra level of indentation imposed by it? I'm thinking about:

```suggestion
    [this_shared=shared_from_this(), this] {
```

Reply:
@rzarzynski's suggestion should work, I think.